tests/AnalyticsEventReporter.test.ts

```diff
@@ -1,34 +1,395 @@
+          clientSdk: {
+            '@yext/analytics': '1.0.0-beta.0'
+          },
+          referrerUrl: 'https://yext.com',
+          count: 5,
+        });
+    });
+
+  it('calling report with no argument and no with call should result in invalid request body and error',
+    async () => {
+      mockPostWithFetch.mockResolvedValue({
+        ok: false,
+        status: 400,
+        statusText: 'Bad Request',
+        json: jest.fn().mockResolvedValue({}),
+        redirected: false,
+        type: 'basic',
+        url: 'https://example.com',
+        clone: jest.fn(),
+        headers: new Headers(),
+        body: null,
+        bodyUsed: false,
+        arrayBuffer: jest.fn(),
+        blob: jest.fn(),
+        formData: jest.fn(),
+        text: jest.fn()
+      });
+
+      const navigator = { userAgent: 'Chrome', sendBeacon: () => { return false; }};
+      Object.defineProperty(window, 'navigator', { value: navigator, writable: true});
+      mockPostWithBeacon.mockReturnValue(false);
+
+      const config: AnalyticsConfig = {
+        key: 'validKey',
+        region: RegionEnum.EU,
+        forceFetch: true,
+      };
+
+      const service: AnalyticsEventService = new AnalyticsEventReporter(config);
+
+      try {
+        await service.report({});
+        // Fail test if above expression doesn't throw anything.
+        expect(true).toBe(false);
+      } catch (e) {
+        expect(e.message).toBe('Events API responded with 400: Bad Request');
+      }
+
+      /** Expect merge to have completed correctly, but the request
+       * body to be invalid as action was never added **/
+      expect(mockPostWithFetch).toHaveBeenCalledWith(
+        'https://eu.yextevents.com/accounts/me/events',
+        {
+          authorization: 'KEY validKey',
+          clientSdk: {
+            '@yext/analytics': '1.0.0-beta.0'
+          },
+        });
+    });
+
+  it('calling with more than once should result in unique objects that can be called with unique values',
+    async () => {
+      mockPostWithBeacon.mockReturnValueOnce(true);
+      mockUseBeacon.mockReturnValueOnce(true);
+      const navigator = { userAgent: 'Firefox', sendBeacon: () => { return true; }};
+      Object.defineProperty(window, 'navigator', { value: navigator, writable: true});
+
+      const config: AnalyticsConfig = {
+        key: 'validKey',
+        region: RegionEnum.EU,
+        forceFetch: false,
+      };
+
+      const reporter = new AnalyticsEventReporter(config).with({
+        action: 'ADD_TO_CART',
+        referrerUrl: 'https://yext.com',
+        count: 5,
+      });
+
+      const unaffectedReporter = JSON.parse(JSON.stringify(reporter));
+
+      const reporter2 = reporter.with({
+        action: 'APPLY',
+        authorization: 'Bearer shouldNotUpdate',
+        destinationUrl: 'https://google.com',
+        clientSdk: {
+          chat: '1.0.1.0',
+        }
+      });
+
+      const res = await reporter2.report();
```

Comment:
Let's have another report() call for reporter1, and verify that the results it's been called with.